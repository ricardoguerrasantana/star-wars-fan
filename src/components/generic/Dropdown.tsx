// React
import { ReactNode, useEffect, useState } from 'react';

// debugger
import Debug from "debug";
const log = Debug("App:Dropdown");
log.log = console.log.bind(console);

// Prop types
export type Props = {
  bodyComponent: ReactNode;
  dropdownStyles: {
    body: string;
    container: string;
    header: string;
  };
  headerComponent: ReactNode;
  id: string;
  displayedDropdown?: string;
  setDisplayedDropdown?: (id: string) => void;
}

/**  Dropdown generic component is compound of a header and a body 
 * boxes which should receive customized components. By default the 
 * body is hidden. When clicking the header it displays the body and 
 * hide it if clicked again. This Dropdown may transfer control to 
 * hide the body if displayedDropdown and setDisplayedDropdown props 
 * are defined from parent component as state managers. */
function Dropdown({ bodyComponent, dropdownStyles, headerComponent, id, displayedDropdown , setDisplayedDropdown }: Props) {
  log('Rendering...');

  // Internal state to determine whether dropdown body is displayed or not
  const [displayBody , setDisplayBody] = useState<boolean>(false);
  
  useEffect(() => {
    // Closes dropdown body if external control has been set up.
    if (displayedDropdown && displayedDropdown !== id) {
      setDisplayBody(false);
    }
  }, [displayedDropdown, id]);

  function handleClick(): void {
    if (setDisplayedDropdown) {
      /** Identifies the displayed dropdown so parent component can target it 
       * if external control has been set up. */
      setDisplayedDropdown(id);
    }
    setDisplayBody(prevDisplayBody => !prevDisplayBody);
  }

  return (
    <div className={dropdownStyles.container} >
      <div 
        className={dropdownStyles.header}
        onClick={handleClick}
      >
        {headerComponent}
      </div>
      <div 
        className={dropdownStyles.body}
        style={displayBody ? undefined : {display: "none"}} 
      >
        {bodyComponent}
      </div>
    </div>
  );
}

export default Dropdown;