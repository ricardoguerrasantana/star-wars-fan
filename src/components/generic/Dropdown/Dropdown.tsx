// React
import { ReactNode, useEffect, useState } from 'react';

// debugger
import Debug from "debug";
const log = Debug("App:Dropdown");
log.log = console.log.bind(console);

// Prop types
export type Props = {
  bodyComponent: ReactNode;
  headerComponent: ReactNode;
  id: string;
  styles: {
    body: string;
    container: string;
    header: string;
  };
  /** Set up when it is necessary to collapse 
   * the Dropdown body from outside. */
  selectedDropdownId?: string;
  setSelectedDropdownId?: (id: string) => void;
}

/**  Dropdown generic component is compound of a header and a body 
 * boxes which should receive customized components. By default 
 * the body is hidden. When clicking the header it displays the 
 * body and hide it if clicked again. This Dropdown may transfer 
 * control to hide the body if selectedDropdownId and 
 * setSelectedDropdownId props are defined from parent component 
 * as state managers. */
function Dropdown({ bodyComponent, styles, headerComponent, id, selectedDropdownId , setSelectedDropdownId }: Props) {
  log("Rendering...");

  /** Controls whether the body is displayed or not. */
  const [displayBody , setDisplayBody] = useState<boolean>(false);
  
  /** Controls whether body should be open or closed when 
   * external control has been set up. */
  useEffect(() => {
    if (selectedDropdownId) {
      setDisplayBody(selectedDropdownId === id ? true : false);
    }
  }, [selectedDropdownId, id]);

  function handleClick(): void {
    if (setSelectedDropdownId) {
      /** External control has been set up. 
       * Identifies this Dropdown in parent component. */
      setSelectedDropdownId(id);
    }
    setDisplayBody(prevDisplayBody => !prevDisplayBody);
  }

  return (
    <div 
      className={styles.container} 
      data-testid="Dropdown"
    >
      <div 
        className={styles.header}
        data-testid="Header"
        onClick={handleClick}
      >
        {headerComponent}
      </div>
      <div 
        className={styles.body}
        data-testid="Body"
        style={displayBody ? undefined : {display: "none"}} 
      >
        {bodyComponent}
      </div>
    </div>
  );
}

export default Dropdown;