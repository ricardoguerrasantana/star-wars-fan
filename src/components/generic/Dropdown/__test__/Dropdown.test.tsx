// Testing libraries
import { render, screen, fireEvent, within } from '@testing-library/react';
import renderer from "react-test-renderer";

// Component
import Dropdown from '../Dropdown';

describe("<Dropdown />", () => {
  /** Fake props to be passed down to <Dropdown /> */
  const bodyComponent = (
    <div>
      <p>Body</p>
      <button type="button">Back</button>
      <button type="button">Next</button>
    </div>
  );
  
  const headerComponent = (<h1>Tile</h1>);

  const styles = {
    body: "body",
    container: "container",
    header: "header",
  };

  const MockDropdown = () => (
    <Dropdown 
      bodyComponent={bodyComponent}
      headerComponent={headerComponent}
      id="Dropdown"
      styles={styles}
    />
  );
      
  it("should render Dropdown component", () => {
    render(<MockDropdown />);
    // screen.debug();
    
    expect(screen.getByTestId("Dropdown")).toBeInTheDocument();
  });
    
  it(`should contain header and body components.`, () => {
    const { getByTestId } = render(<MockDropdown />);
    const Dropdown = getByTestId("Dropdown");
    // Checks whether the expected elements were inserted.
    expect(within(Dropdown).getByTestId("Header")).toBeTruthy();
    expect(within(Dropdown).getByTestId("Body")).toBeTruthy();
  });
    
  it(`clicking on header should be able to display body when closed or collapse body when open.`, () => {
    const { getByTestId } = render(<MockDropdown />);
    const Dropdown = getByTestId("Dropdown");
    const header = within(Dropdown).getByTestId("Header");
    const body = within(Dropdown).getByTestId("Body");
    /** Mocks clicking on header to display body. */
    fireEvent.click(header);
    /** body should be displayed. */
    expect(body).not.toHaveStyle("display: none;");
    /** Mocks clicking on header to collapse body. */
    fireEvent.click(header);
    /** body should Not be displayed. */
    expect(body).toHaveStyle("display: none;");
    /** Mocks clicking on header to display body once again. */
    fireEvent.click(header);
    /** Finally, body should be displayed. */
    expect(body).not.toHaveStyle("display: none;");
  });

  it(`should match the snapshot of initial render`, () => {
      const tree = renderer.create(<MockDropdown />, ).toJSON();
      expect(tree).toMatchSnapshot();
  });
});