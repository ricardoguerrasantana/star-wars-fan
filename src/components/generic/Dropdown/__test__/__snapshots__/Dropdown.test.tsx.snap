// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> should match the snapshot of initial render 1`] = `
<div
  className="container"
  data-testid="Dropdown"
>
  <div
    className="header"
    data-testid="Header"
    onClick={[Function]}
  >
    <h1>
      Tile
    </h1>
  </div>
  <div
    className="body"
    data-testid="Body"
    style={
      Object {
        "display": "none",
      }
    }
  >
    <div>
      <p>
        Body
      </p>
      <button
        type="button"
      >
        Back
      </button>
      <button
        type="button"
      >
        Next
      </button>
    </div>
  </div>
</div>
`;
